File: src/component/ToggleButton.java
package component;

import org.newdawn.slick.gui.GUIContext;

/**
 * Toggle Button is a button that holds its state until its clicked again.
 */
public class ToggleButton extends Button {
	private boolean disableAutoToggle;
	
	/**
	 * Creates a {@code ToggleButton} with a label, width, and height.
	 * @param container The GUI context
	 * @param width The width
	 * @param height The height
	 * @param label A label
	 */
	public ToggleButton(GUIContext context, int width, int height, Label label) {
		super(context, width, height, label);
		
		disableAutoToggle = false;
	}
	
	@Override
	public void mousePressed(int button, int mx, int my) {
		if (!isVisible() || !isAcceptingInput()) {
			return;
		}
		
		if (button == 0 && isMouseOver() && !isDisabled()) {
			if (!disableAutoToggle) {
				setActive(!isActive());
			} else {
				setActive(true);
			}
			
			input.consumeEvent();
		}
	}
	
	@Override
	public void mouseReleased(int button, int mx, int my) {
		if (!isVisible()) {
			return;
		}
		
 		if (button == 0 && isMouseOver() && !isDisabled() && isActive()) {
			if (disableAutoToggle) {
				setActive(false);
			}
			
			notifyListeners();
			input.consumeEvent();
		}
	}
	
	/**
	 * Change the value to disable auto toggle.
	 * @param disableAutoToggle Whether auto toggle will be on or off
	 */
	public void setDisableAutoToggle(boolean disableAutoToggle) {
		this.disableAutoToggle = disableAutoToggle;
	}
	
	/**
	 * Get the disable auto toggle status.
	 * @return Auto toggle status
	 */
	public boolean getDisableAutoToggle() {
		return disableAutoToggle;
	}
}

File: web/src/component/ToggleButton.ts
import { Button } from './Button';
import { Label } from './Label';

export class ToggleButton extends Button {
  private disableAutoToggle: boolean = false;

  constructor(width: number, height: number, label: Label) {
    super(width, height, label);
  }

  override mousePressed(button: number, _mx: number, _my: number): void {
    if (!this.isVisible() || !this.isAcceptingInput()) {
      return;
    }

    if (button === 0 && this.isMouseOver() && !this.isDisabled()) {
      if (!this.disableAutoToggle) {
        this.setActive(!this.isActive());
      } else {
        this.setActive(true);
      }
    }
  }

  override mouseReleased(button: number, _mx: number, _my: number): void {
    if (!this.isVisible()) {
      return;
    }

    if (button === 0 && this.isMouseOver() && !this.isDisabled() && this.isActive()) {
      if (this.disableAutoToggle) {
        this.setActive(false);
      }

      this.notifyListeners();
    }
  }

  setDisableAutoToggle(disableAutoToggle: boolean): void {
    this.disableAutoToggle = disableAutoToggle;
  }

  getDisableAutoToggle(): boolean {
    return this.disableAutoToggle;
  }
}
